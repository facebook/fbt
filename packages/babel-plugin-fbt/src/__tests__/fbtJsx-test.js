/**
 * (c) Meta Platforms, Inc. and affiliates. Confidential and proprietary.
 *
 * @emails oncall+i18n_fbt_js
 * @format
 */

/* eslint-disable fb-www/gender-neutral-language */

jest.autoMockOff();

const {
  jsCodeFbtCallSerializer,
  snapshotTransformKeepJsx,
  withFbtRequireStatement,
} = require('./FbtTestUtil');
const {TestUtil} = require('fb-babel-plugin-utils');

expect.addSnapshotSerializer(jsCodeFbtCallSerializer);

const testData = {
  'should convert simple strings': {
    input: withFbtRequireStatement(
      `var x = <fbt desc="It's simple">A simple string</fbt>;`,
    ),
  },

  'should filter comment and empty expressions from children': {
    input: withFbtRequireStatement(
      `var x = <fbt desc="It's simple">
        {
        }A sim{/*
          ignore
          me
          */}ple s{ }tri{}ng{/*ignore me*/}</fbt>;`,
    ),
  },

  'should strip out newlines': {
    input: withFbtRequireStatement(
      `var x =
        <fbt desc="Test trailing space when not last child">
          Preamble
          <fbt:param name="parm">{blah}</fbt:param>
        </fbt>;
      baz();`,
    ),
  },

  'should strip out newlines in Reactish <Fbt>': {
    input: withFbtRequireStatement(
      `var x =
        <Fbt desc="Test trailing space when not last child">
          Preamble <FbtParam name="parm">{blah}</FbtParam>
        </Fbt>;
      baz();`,
    ),
  },

  'should strip out more newlines': {
    input: withFbtRequireStatement(
      `var x =
        <fbt desc="moar lines">
          A simple string...
          with some other stuff.
        </fbt>;
        baz();`,
    ),
  },

  'Squelch whitespace when in an expression': {
    input: withFbtRequireStatement(
      `var x =
        <fbt desc="squelched">
          {"Squelched white space... "}
          with some
          {' other stuff.'}
        </fbt>;
        baz();`,
    ),
  },

  'Enable explicit whitespace': {
    input: withFbtRequireStatement(
      `var x = <fbt desc="squelched">
        <fbt:param name="one">{one}</fbt:param>
        {" "}
        <fbt:param name="two">{two}</fbt:param>
        {\` \`}
        <fbt:param name="three">{three}</fbt:param>
      </fbt>;`,
    ),
  },

  'should handle params': {
    input: withFbtRequireStatement(
      `var x = <fbt desc="a message!">
          A parameterized message to:
          <fbt:param name="personName">{theName}</fbt:param>
        </fbt>;`,
    ),
  },

  'should handle empty string': {
    input: withFbtRequireStatement(
      `var x = <fbt desc="a message!">
        A parameterized message to:
        <fbt:param name="emptyString"> </fbt:param>
      </fbt>;`,
    ),
  },

  'should handle concatenated descriptions': {
    input: withFbtRequireStatement(
      `<fbt desc={"A very long description " + "that we will concatenate " +
        "a few times"}
        project={"With" + "a" + "project"}>
        Here it is
      </fbt>;`,
    ),
  },

  'should handle template descriptions': {
    input: withFbtRequireStatement(
      `<fbt desc={\`A very long description
        that will be a
        template across multiple lines\`}
        project={"With" + "a" + "project"}>
        Here it is
      </fbt>;`,
    ),
  },

  'should be able to nest within React nodes': {
    input: withFbtRequireStatement(
      `var x = <div>
        <fbt desc="nested!">
          A nested string
        </fbt>
      </div>;`,
    ),
  },

  'should be able to house arbitrary markup within fbt:param nodes': {
    input: withFbtRequireStatement(
      `<div>
        <fbt desc="...">
          <fbt:param name="time">{formatDate(date, "F d, Y")}</fbt:param>
           by
          <fbt:param name="user name">
            <Link href={{url:user.link}}>
              {user.name}
            </Link>
          </fbt:param>
        </fbt>
      </div>;`,
    ),
  },

  'should handle enums (with array values)': {
    input: withFbtRequireStatement(
      `var x = <fbt desc="enums!">
        Click to see
        <fbt:enum
          enum-range={[
            "groups",
            "photos",
            "videos"
          ]}
          value={id}
        />
      </fbt>;`,
    ),
  },

  'should handle enums (with object values)': {
    input: withFbtRequireStatement(
      `var x = <fbt desc="enums!">
        Click to see
        <fbt:enum
          enum-range={{
            id1: "groups",
            id2: "photos",
            id3: "videos"
          }}
          value={id}
        />
      </fbt>;`,
    ),
  },

  'should handle enums with more text': {
    input: withFbtRequireStatement(
      `var x = <fbt desc="enums!">
        Click to see
        <fbt:enum
          enum-range={{
            id1: "groups",
            id2: "photos",
            id3: "videos"
          }}
          value={id}
        />
        Hey-hey!
      </fbt>;`,
    ),
  },

  'should handle variations': {
    input: withFbtRequireStatement(
      `var x = <fbt desc="variations!">
        Click to see
        <fbt:param name="count" number="true">{c}</fbt:param>
        links
      </fbt>;`,
    ),
  },

  'should handle number={true} - (same output as above test)': {
    input: withFbtRequireStatement(
      `var x = <fbt desc="variations!">
        Click to see
        <fbt:param name="count" number={true}>{c}</fbt:param>
        links
      </fbt>;`,
    ),
  },

  'should correctly destruct expression values in options': {
    input: withFbtRequireStatement(
      `<fbt desc="d">str
        <fbt:param name="count" number={someNum}>
          {getNum()}
        </fbt:param>
      </fbt>`,
    ),
  },

  'should insert param value for same-param': {
    input: withFbtRequireStatement(
      `<fbt desc="d">str
        <fbt:param name="foo">{Bar}</fbt:param> and
        <fbt:same-param name="foo"/>
      </fbt>`,
    ),
  },

  'should treat multiline descs as a single line': {
    input: withFbtRequireStatement(
      `<fbt desc="hi how are you today im doing well i guess
        how is your mother is she well yeah why not lets go
        home and never come back.">
        lol
      </fbt>`,
    ),
  },

  'should not insert extra space': {
    input: withFbtRequireStatement(
      `<fbt desc="Greating in i18n demo">
        Hello, <fbt:param name="guest">
          {guest}
        </fbt:param>!
      </fbt>`,
    ),
  },

  'should handle single expression with concentated strings': {
    input: withFbtRequireStatement(
      `<fbt desc="foo">
        {"foo" + "bar"}
      </fbt>`,
    ),
  },

  'should throw on invalid attributes in fbt:param': {
    input: withFbtRequireStatement(
      `<fbt desc="some-desc">
        <fbt:param name="foo" qux="foo" desc="foo-desc">
          {foo}
        </fbt:param>
      </fbt>`,
    ),

    throws: `Invalid option "qux". Only allowed: number, gender, name`,
  },

  'should ignore __private attributes': {
    input: withFbtRequireStatement(
      `<fbt __self="fbt" desc="some-desc">
        <fbt:param __self="param" name="foo">
          {foo}
        </fbt:param>
      </fbt>`,
    ),
  },

  'should ignore non-expression children in fbt:param': {
    input: withFbtRequireStatement(
      `<fbt desc="some-desc">
        <fbt:param name="foo">
          !{foo}!
        </fbt:param>
      </fbt>`,
    ),
  },

  'should maintain order of params and enums': {
    input: withFbtRequireStatement(
      `<fbt desc="some-desc">
        Hello,
        <fbt:param name="foo">
          {foo}
        </fbt:param>
        <fbt:enum enum-range={["x", "y"]} value={x} />
        <fbt:param name="bar" number={n}>
          {bar}
        </fbt:param>
      </fbt>`,
    ),
  },

  'should support html escapes': {
    input: withFbtRequireStatement(
      `<fbt desc="foo &quot;bar&quot;">&times;</fbt>`,
    ),
  },

  'should support unicode characters': {
    input: withFbtRequireStatement(
      `// A backslash \\ in comments
      <fbt desc="unicode characters">
        A copyright sign {'\\u00A9'},
        a multi byte character {'\\uD83D\\uDCA9'},
        and a backslash {'\\\\'}.
      </fbt>`,
    ),
  },

  'should handle object pronoun': {
    input: withFbtRequireStatement(
      `<fbt desc={"d"} project={"p"}>
          I know <fbt:pronoun type="object" gender={gender}/>.
        </fbt>;`,
    ),
  },

  'should handle object pronoun (react native)': {
    input: withFbtRequireStatement(
      `<fbt desc={"d"} project={"p"}>
          I know <fbt:pronoun type="object" gender={gender}/>.
        </fbt>;`,
    ),

    options: {
      reactNativeMode: true,
    },
  },

  'should handle subject+reflexive pronouns': {
    input:
      // She wished herself a happy birthday.
      withFbtRequireStatement(
        `<fbt desc={"d"} project={"p"}>
          <fbt:pronoun type="subject" gender={gender} capitalize={true} human={true}/>
          wished <fbt:pronoun type="reflexive" gender={gender} human={true}/> a happy birthday.
        </fbt>;`,
      ),
  },

  'should handle subject+reflexive pronouns (react native)': {
    input:
      // She wished herself a happy birthday.
      withFbtRequireStatement(
        `<fbt desc={"d"} project={"p"}>
          <fbt:pronoun type="subject" gender={gender} capitalize={true} human={true}/>
          wished
          <fbt:pronoun type="reflexive" gender={gender} human={true}/>
          a happy birthday.
        </fbt>;`,
      ),

    options: {
      reactNativeMode: true,
    },
  },

  'fbt:param with multiple children should error': {
    input: withFbtRequireStatement(
      `<fbt desc="some-desc">
        <fbt:param name="foo">
          {foo}
          {bar}
        </fbt:param>
      </fbt>`,
    ),

    throws: `fbt:param expects an {expression} or JSX element, and only one`,
  },

  'fbt:param with multiple empty expression containers should be ok': {
    input: withFbtRequireStatement(
      `<fbt desc="some-desc">
        <fbt:param name="foo">
          {}
          {/* comment */}
          {foo}
          {}
        </fbt:param>
      </fbt>`,
    ),
  },
};

describe('Test declarative (jsx) fbt syntax translation', () =>
  TestUtil.testSection(testData, snapshotTransformKeepJsx, {
    matchSnapshot: true,
  }));

describe('Test fbt transforms without the jsx transform', () => {
  it('not nested', () => {
    expect(
      snapshotTransformKeepJsx(`
        const fbt = require("fbt");
        let x =
          <fbt desc="nested!">
            A nested string
          </fbt>;
      `),
    ).toMatchSnapshot(); // Should be like fbt._()
  });

  it('nested in div', () => {
    expect(
      snapshotTransformKeepJsx(`
        const fbt = require("fbt");
        let x =
          <div>
            <fbt desc="nested!">
              A nested string
            </fbt>
          </div>;
      `),
    ).toMatchSnapshot(); // Should be like <div>{fbt._()}</div>
  });

  it('short bool syntax for doNotExtract attribute', () => {
    expect(
      snapshotTransformKeepJsx(`
        const fbt = require("fbt");
        let x = <fbt desc="" doNotExtract>Test</fbt>;
      `),
    ).toMatchSnapshot();
  });

  it('short bool syntax for number attribute', () => {
    expect(
      snapshotTransformKeepJsx(`
        const fbt = require("fbt");
        let x =
          <fbt desc="">
            <fbt:param name="name" number>{'name'}</fbt:param>
          </fbt>;
      `),
    ).toMatchSnapshot();
  });

  describe('when using within template literals', () => {
    it('should work with a basic <fbt>', () => {
      expect(
        snapshotTransformKeepJsx(`
          const fbt = require("fbt");
          html\`<div>
            \${
              <fbt desc="some desc" project="some project">
                basic text
              </fbt>
            }
          </div>\`;
        `),
      ).toMatchSnapshot();
    });

    it('should work with basic <fbt> auto-parameterization', () => {
      expect(
        snapshotTransformKeepJsx(`
          const fbt = require("fbt");
          html\`<div>
            \${
              <fbt desc="some desc" project="some project">
                outer text
                <strong>
                  bold text
                </strong>
              </fbt>
            }
          </div>\`;
        `),
      ).toMatchSnapshot();
    });

    it('should dedupe plurals', () => {
      expect(
        snapshotTransformKeepJsx(`
          const fbt = require("fbt");
          <fbt desc="desc...">
            There
            <fbt:plural count={num} many="are">is</fbt:plural>{' '}
            <fbt:plural count={num} showCount="yes" value={hi()}>
              photo
            </fbt:plural>.
          </fbt>
        `),
      ).toMatchSnapshot();
    });

    it('should work with a nested <fbt> within an <fbt:param>', () => {
      expect(
        snapshotTransformKeepJsx(`
          const fbt = require("fbt");
          html\`<div>
            \${
              <fbt desc="some desc" project="some project">
                outer text
                <fbt:param name="param text">
                  {
                    html\`<strong>
                      \${
                        <fbt desc="inner string">
                          inner text
                          <fbt:param name="inner param">
                            {'bold'}
                          </fbt:param>
                        </fbt>
                      }
                    </strong>\`
                  }
                </fbt:param>
              </fbt>
            }
          </div>\`;
        `),
      ).toMatchSnapshot();
    });
  });

  // TODO(T94644387) fix preserving whitespace for JSX text
  it('should fail to preserve whitespace in text when preserveWhitespace=true (known bug)', () => {
    expect(
      snapshotTransformKeepJsx(`
        const fbt = require('fbt');
        <fbt desc="desc with 3   spaces" preserveWhitespace={true}>
          Some text with 3   spaces in between.
        </fbt>;
      `),
    ).toMatchSnapshot();
  });

  // TODO(T78914132): actually, we should NOT insert a space between two <fbt:plural>
  // that don't neighbor raw text to match Hack fbt parity.
  // But there's not much point fixing this before the fbt autoparam work.
  // See Hack fbt equivalent: https://fburl.com/intl/zkacwqtj
  // See also JS fbt fiddle: https://fburl.com/intl/ha5dryng
  it(`[legacy buggy behavior] <fbt:pronoun> should insert a space character between two fbt constructs that don't neighbor raw text`, () =>
    expect(
      snapshotTransformKeepJsx(`
        const fbt = require("fbt");
        <fbt desc="">
          You can add
          <fbt:plural count={count} many="these">
            this
          </fbt:plural>
          <fbt:plural count={count} many="tags">
            tag
          </fbt:plural>
          to anything.
        </fbt>
      `),
    ).toMatchSnapshot());
});
