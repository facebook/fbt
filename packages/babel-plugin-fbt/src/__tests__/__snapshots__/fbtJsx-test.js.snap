// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test declarative (jsx) fbt syntax translation Enable explicit whitespace 1`] = `
const fbt = require("fbt");

var x = fbt._(
  /* __FBT__ start */ {
    jsfbt: { t: { desc: "squelched", text: "{one} {two} {three}" }, m: [] },
    project: "",
  } /* __FBT__ end */,
  [fbt._param("one", one), fbt._param("two", two), fbt._param("three", three)]
);

`;

exports[`Test declarative (jsx) fbt syntax translation Squelch whitespace when in an expression 1`] = `
const fbt = require("fbt");

var x = fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      t: {
        desc: "squelched",
        text: "Squelched white space... with some other stuff.",
      },
      m: [],
    },
    project: "",
  } /* __FBT__ end */
);

baz();

`;

exports[`Test declarative (jsx) fbt syntax translation fbt:param with multiple empty expression containers should be ok 1`] = `
const fbt = require("fbt");

fbt._(
  /* __FBT__ start */ {
    jsfbt: { t: { desc: "some-desc", text: "{foo}" }, m: [] },
    project: "",
  } /* __FBT__ end */,
  [fbt._param("foo", foo)]
);

`;

exports[`Test declarative (jsx) fbt syntax translation should be able to house arbitrary markup within fbt:param nodes 1`] = `
const fbt = require("fbt");

<div>
  {fbt._(
    /* __FBT__ start */ {
      jsfbt: { t: { desc: "...", text: "{time} by {user name}" }, m: [] },
      project: "",
    } /* __FBT__ end */,
    [
      fbt._param("time", formatDate(date, "F d, Y")),
      fbt._param(
        "user name",
        <Link
          href={{
            url: user.link,
          }}
        >
          {user.name}
        </Link>
      ),
    ]
  )}
</div>;

`;

exports[`Test declarative (jsx) fbt syntax translation should be able to nest within React nodes 1`] = `
const fbt = require("fbt");

var x = (
  <div>
    {fbt._(
      /* __FBT__ start */ {
        jsfbt: { t: { desc: "nested!", text: "A nested string" }, m: [] },
        project: "",
      } /* __FBT__ end */
    )}
  </div>
);

`;

exports[`Test declarative (jsx) fbt syntax translation should convert simple strings 1`] = `
const fbt = require("fbt");

var x = fbt._(
  /* __FBT__ start */ {
    jsfbt: { t: { desc: "It's simple", text: "A simple string" }, m: [] },
    project: "",
  } /* __FBT__ end */
);

`;

exports[`Test declarative (jsx) fbt syntax translation should correctly destruct expression values in options 1`] = `
const fbt = require("fbt");

fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      t: { "*": { desc: "d", text: "str {count}" } },
      m: [{ token: "count", type: 2 }],
    },
    project: "",
  } /* __FBT__ end */,
  [fbt._param("count", getNum(), [0, someNum])]
);

`;

exports[`Test declarative (jsx) fbt syntax translation should filter comment and empty expressions from children 1`] = `
const fbt = require("fbt");

var x = fbt._(
  /* __FBT__ start */ {
    jsfbt: { t: { desc: "It's simple", text: "A simple string" }, m: [] },
    project: "",
  } /* __FBT__ end */
);

`;

exports[`Test declarative (jsx) fbt syntax translation should handle concatenated descriptions 1`] = `
const fbt = require("fbt");

fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      t: {
        desc: "A very long description that we will concatenate a few times",
        text: "Here it is",
      },
      m: [],
    },
    project: "Withaproject",
  } /* __FBT__ end */
);

`;

exports[`Test declarative (jsx) fbt syntax translation should handle empty string 1`] = `
const fbt = require("fbt");

var x = fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      t: {
        desc: "a message!",
        text: "A parameterized message to: {emptyString}",
      },
      m: [],
    },
    project: "",
  } /* __FBT__ end */,
  [fbt._param("emptyString", " ")]
);

`;

exports[`Test declarative (jsx) fbt syntax translation should handle enums (with array values) 1`] = `
const fbt = require("fbt");

var x = fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      t: {
        groups: { desc: "enums!", text: "Click to see groups" },
        photos: { desc: "enums!", text: "Click to see photos" },
        videos: { desc: "enums!", text: "Click to see videos" },
      },
      m: [null],
    },
    project: "",
  } /* __FBT__ end */,
  [
    fbt._enum(id, {
      groups: "groups",
      photos: "photos",
      videos: "videos",
    }),
  ]
);

`;

exports[`Test declarative (jsx) fbt syntax translation should handle enums (with object values) 1`] = `
const fbt = require("fbt");

var x = fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      t: {
        id1: { desc: "enums!", text: "Click to see groups" },
        id2: { desc: "enums!", text: "Click to see photos" },
        id3: { desc: "enums!", text: "Click to see videos" },
      },
      m: [null],
    },
    project: "",
  } /* __FBT__ end */,
  [
    fbt._enum(id, {
      id1: "groups",
      id2: "photos",
      id3: "videos",
    }),
  ]
);

`;

exports[`Test declarative (jsx) fbt syntax translation should handle enums with more text 1`] = `
const fbt = require("fbt");

var x = fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      t: {
        id1: { desc: "enums!", text: "Click to see groups Hey-hey!" },
        id2: { desc: "enums!", text: "Click to see photos Hey-hey!" },
        id3: { desc: "enums!", text: "Click to see videos Hey-hey!" },
      },
      m: [null],
    },
    project: "",
  } /* __FBT__ end */,
  [
    fbt._enum(id, {
      id1: "groups",
      id2: "photos",
      id3: "videos",
    }),
  ]
);

`;

exports[`Test declarative (jsx) fbt syntax translation should handle number={true} - (same output as above test) 1`] = `
const fbt = require("fbt");

var x = fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      t: { "*": { desc: "variations!", text: "Click to see {count} links" } },
      m: [{ token: "count", type: 2 }],
    },
    project: "",
  } /* __FBT__ end */,
  [fbt._param("count", c, [0])]
);

`;

exports[`Test declarative (jsx) fbt syntax translation should handle object pronoun (react native) 1`] = `
const fbt = require("fbt");

fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      t: {
        0: { desc: "d", text: "I know this." },
        1: { desc: "d", text: "I know her." },
        2: { desc: "d", text: "I know him." },
        "*": { desc: "d", text: "I know them." },
      },
      m: [{ type: 3 }],
    },
    project: "p",
  } /* __FBT__ end */,
  [fbt._pronoun(0, gender)]
);

`;

exports[`Test declarative (jsx) fbt syntax translation should handle object pronoun 1`] = `
const fbt = require("fbt");

fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      t: {
        0: { desc: "d", text: "I know this." },
        1: { desc: "d", text: "I know her." },
        2: { desc: "d", text: "I know him." },
        "*": { desc: "d", text: "I know them." },
      },
      m: [null],
    },
    project: "p",
  } /* __FBT__ end */,
  [fbt._pronoun(0, gender)]
);

`;

exports[`Test declarative (jsx) fbt syntax translation should handle params 1`] = `
const fbt = require("fbt");

var x = fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      t: {
        desc: "a message!",
        text: "A parameterized message to: {personName}",
      },
      m: [],
    },
    project: "",
  } /* __FBT__ end */,
  [fbt._param("personName", theName)]
);

`;

exports[`Test declarative (jsx) fbt syntax translation should handle single expression with concentated strings 1`] = `
const fbt = require("fbt");

fbt._(
  /* __FBT__ start */ {
    jsfbt: { t: { desc: "foo", text: "foobar" }, m: [] },
    project: "",
  } /* __FBT__ end */
);

`;

exports[`Test declarative (jsx) fbt syntax translation should handle subject+reflexive pronouns (react native) 1`] = `
const fbt = require("fbt");

fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      t: {
        1: { 1: { desc: "d", text: "She wished herself a happy birthday." } },
        2: { 2: { desc: "d", text: "He wished himself a happy birthday." } },
        "*": {
          "*": { desc: "d", text: "They wished themselves a happy birthday." },
        },
      },
      m: [{ type: 3 }, { type: 3 }],
    },
    project: "p",
  } /* __FBT__ end */,
  [
    fbt._pronoun(3, gender, {
      human: 1,
    }),
    fbt._pronoun(2, gender, {
      human: 1,
    }),
  ]
);

`;

exports[`Test declarative (jsx) fbt syntax translation should handle subject+reflexive pronouns 1`] = `
const fbt = require("fbt");

fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      t: {
        1: { 1: { desc: "d", text: "She wished herself a happy birthday." } },
        2: { 2: { desc: "d", text: "He wished himself a happy birthday." } },
        "*": {
          "*": { desc: "d", text: "They wished themselves a happy birthday." },
        },
      },
      m: [null, null],
    },
    project: "p",
  } /* __FBT__ end */,
  [
    fbt._pronoun(3, gender, {
      human: 1,
    }),
    fbt._pronoun(2, gender, {
      human: 1,
    }),
  ]
);

`;

exports[`Test declarative (jsx) fbt syntax translation should handle template descriptions 1`] = `
const fbt = require("fbt");

fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      t: {
        desc:
          "A very long description that will be a template across multiple lines",
        text: "Here it is",
      },
      m: [],
    },
    project: "Withaproject",
  } /* __FBT__ end */
);

`;

exports[`Test declarative (jsx) fbt syntax translation should handle variations 1`] = `
const fbt = require("fbt");

var x = fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      t: { "*": { desc: "variations!", text: "Click to see {count} links" } },
      m: [{ token: "count", type: 2 }],
    },
    project: "",
  } /* __FBT__ end */,
  [fbt._param("count", c, [0])]
);

`;

exports[`Test declarative (jsx) fbt syntax translation should ignore __private attributes 1`] = `
const fbt = require("fbt");

fbt._(
  /* __FBT__ start */ {
    jsfbt: { t: { desc: "some-desc", text: "{foo}" }, m: [] },
    project: "",
  } /* __FBT__ end */,
  [fbt._param("foo", foo)]
);

`;

exports[`Test declarative (jsx) fbt syntax translation should ignore non-expression children in fbt:param 1`] = `
const fbt = require("fbt");

fbt._(
  /* __FBT__ start */ {
    jsfbt: { t: { desc: "some-desc", text: "{foo}" }, m: [] },
    project: "",
  } /* __FBT__ end */,
  [fbt._param("foo", foo)]
);

`;

exports[`Test declarative (jsx) fbt syntax translation should insert param value for same-param 1`] = `
const fbt = require("fbt");

fbt._(
  /* __FBT__ start */ {
    jsfbt: { t: { desc: "d", text: "str {foo} and {foo}" }, m: [] },
    project: "",
  } /* __FBT__ end */,
  [fbt._param("foo", Bar)]
);

`;

exports[`Test declarative (jsx) fbt syntax translation should maintain order of params and enums 1`] = `
const fbt = require("fbt");

fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      t: {
        x: { "*": { desc: "some-desc", text: "Hello, {foo}x{bar}" } },
        y: { "*": { desc: "some-desc", text: "Hello, {foo}y{bar}" } },
      },
      m: [null, { token: "bar", type: 2 }],
    },
    project: "",
  } /* __FBT__ end */,
  [
    fbt._enum(x, {
      x: "x",
      y: "y",
    }),
    fbt._param("bar", bar, [0, n]),
    fbt._param("foo", foo),
  ]
);

`;

exports[`Test declarative (jsx) fbt syntax translation should not insert extra space 1`] = `
const fbt = require("fbt");

fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      t: { desc: "Greating in i18n demo", text: "Hello, {guest}!" },
      m: [],
    },
    project: "",
  } /* __FBT__ end */,
  [fbt._param("guest", guest)]
);

`;

exports[`Test declarative (jsx) fbt syntax translation should strip out more newlines 1`] = `
const fbt = require("fbt");

var x = fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      t: {
        desc: "moar lines",
        text: "A simple string... with some other stuff.",
      },
      m: [],
    },
    project: "",
  } /* __FBT__ end */
);

baz();

`;

exports[`Test declarative (jsx) fbt syntax translation should strip out newlines 1`] = `
const fbt = require("fbt");

var x = fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      t: {
        desc: "Test trailing space when not last child",
        text: "Preamble {parm}",
      },
      m: [],
    },
    project: "",
  } /* __FBT__ end */,
  [fbt._param("parm", blah)]
);

baz();

`;

exports[`Test declarative (jsx) fbt syntax translation should strip out newlines in Reactish <Fbt> 1`] = `
const fbt = require("fbt");

var x = fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      t: {
        desc: "Test trailing space when not last child",
        text: "Preamble {parm}",
      },
      m: [],
    },
    project: "",
  } /* __FBT__ end */,
  [fbt._param("parm", blah)]
);

baz();

`;

exports[`Test declarative (jsx) fbt syntax translation should support html escapes 1`] = `
const fbt = require("fbt");

fbt._(
  /* __FBT__ start */ {
    jsfbt: { t: { desc: 'foo "bar"', text: "×" }, m: [] },
    project: "",
  } /* __FBT__ end */
);

`;

exports[`Test declarative (jsx) fbt syntax translation should support unicode characters 1`] = `
const fbt = require("fbt"); // A backslash \\ in comments

fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      t: {
        desc: "unicode characters",
        text:
          "A copyright sign ©, a multi byte character 💩, and a backslash \\\\.",
      },
      m: [],
    },
    project: "",
  } /* __FBT__ end */
);

`;

exports[`Test declarative (jsx) fbt syntax translation should treat multiline descs as a single line 1`] = `
const fbt = require("fbt");

fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      t: {
        desc:
          "hi how are you today im doing well i guess how is your mother is she well yeah why not lets go home and never come back.",
        text: "lol",
      },
      m: [],
    },
    project: "",
  } /* __FBT__ end */
);

`;

exports[`Test fbt transforms without the jsx transform [legacy buggy behavior] <fbt:pronoun> should insert a space character between two fbt constructs that don't neighbor raw text 1`] = `
const fbt = require("fbt");

fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      t: {
        "*": { "*": { desc: "", text: "You can add thesetags to anything." } },
        _1: { _1: { desc: "", text: "You can add this tag to anything." } },
      },
      m: [null, null],
    },
    project: "",
  } /* __FBT__ end */,
  [fbt._plural(count), fbt._plural(count)]
);

`;

exports[`Test fbt transforms without the jsx transform nested in div 1`] = `
const fbt = require("fbt");

let x = (
  <div>
    {fbt._(
      /* __FBT__ start */ {
        jsfbt: { t: { desc: "nested!", text: "A nested string" }, m: [] },
        project: "",
      } /* __FBT__ end */
    )}
  </div>
);

`;

exports[`Test fbt transforms without the jsx transform not nested 1`] = `
const fbt = require("fbt");

let x = fbt._(
  /* __FBT__ start */ {
    jsfbt: { t: { desc: "nested!", text: "A nested string" }, m: [] },
    project: "",
  } /* __FBT__ end */
);

`;

exports[`Test fbt transforms without the jsx transform short bool syntax for doNotExtract attribute 1`] = `
const fbt = require("fbt");

let x = fbt._(
  /* __FBT__ start */ {
    jsfbt: { t: { desc: "", text: "Test" }, m: [] },
    project: "",
  } /* __FBT__ end */
);

`;

exports[`Test fbt transforms without the jsx transform short bool syntax for number attribute 1`] = `
const fbt = require("fbt");

let x = fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      t: { "*": { desc: "", text: "{name}" } },
      m: [{ token: "name", type: 2 }],
    },
    project: "",
  } /* __FBT__ end */,
  [fbt._param("name", "name", [0])]
);

`;

exports[`Test fbt transforms without the jsx transform should fail to preserve whitespace in text when preserveWhitespace=true (known bug) 1`] = `
const fbt = require("fbt");

fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      t: {
        desc: "desc with 3   spaces",
        text: "Some text with 3 spaces in between.",
      },
      m: [],
    },
    project: "",
  } /* __FBT__ end */
);

`;

exports[`Test fbt transforms without the jsx transform when using within template literals should dedupe plurals 1`] = `
const fbt = require("fbt");

fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      t: {
        "*": { "*": { desc: "desc...", text: "There are {number} photos." } },
        _1: { _1: { desc: "desc...", text: "There is 1 photo." } },
      },
      m: [null, { token: "number", type: 2, singular: true }],
    },
    project: "",
  } /* __FBT__ end */,
  [fbt._plural(num), fbt._plural(num, "number", hi())]
);

`;

exports[`Test fbt transforms without the jsx transform when using within template literals should work with a basic <fbt> 1`] = `
const fbt = require("fbt");

html\`<div>
  \${fbt._(
    /* __FBT__ start */ {
      jsfbt: { t: { desc: "some desc", text: "basic text" }, m: [] },
      project: "some project",
    } /* __FBT__ end */
  )}
</div>\`;

`;

exports[`Test fbt transforms without the jsx transform when using within template literals should work with a nested <fbt> within an <fbt:param> 1`] = `
const fbt = require("fbt");

html\`<div>
  \${fbt._(
    /* __FBT__ start */ {
      jsfbt: {
        t: { desc: "some desc", text: "outer text {param text}" },
        m: [],
      },
      project: "some project",
    } /* __FBT__ end */,
    [
      fbt._param(
        "param text",
        html\`<strong>
          \${fbt._(
            /* __FBT__ start */ {
              jsfbt: {
                t: { desc: "inner string", text: "inner text {inner param}" },
                m: [],
              },
              project: "",
            } /* __FBT__ end */,
            [fbt._param("inner param", "bold")]
          )}
        </strong>\`
      ),
    ]
  )}
</div>\`;

`;

exports[`Test fbt transforms without the jsx transform when using within template literals should work with basic <fbt> auto-parameterization 1`] = `
const fbt = require("fbt");

html\`<div>
  \${fbt._(
    /* __FBT__ start */ {
      jsfbt: {
        t: {
          desc: "some desc",
          text: "outer text {=bold text}",
          tokenAliases: { "=bold text": "=m1" },
        },
        m: [],
      },
      project: "some project",
    } /* __FBT__ end */,
    [
      fbt._implicitParam(
        "=m1",
        <strong>
          {fbt._(
            /* __FBT__ start */ {
              jsfbt: {
                t: {
                  desc: 'In the phrase: "outer text {=bold text}"',
                  text: "bold text",
                },
                m: [],
              },
              project: "some project",
            } /* __FBT__ end */
          )}
        </strong>
      ),
    ]
  )}
</div>\`;

`;
